drivers/volume/portworx/px-statfs/px_statfs.c

```diff
@@ -0,0 +1,180 @@
+{
+    char filename[256] = {0};
+    char major_nr_str[10] = {0};
+    sprintf(filename, "/sys/devices/pxd/%d/major", minor);
+    read_to_str(filename, major_nr_str, 9);
+    if (strlen(major_nr_str) > 0)
+    {
+        int pxd_major = atoi(major_nr_str);
+        if (pxd_major == major)
+        {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/*
+ * For encrypted device
+ * # cat /sys/dev/block/253:0/dm/uuid
+ *    CRYPT-LUKS1-41d2686ad84f43d2a64f20502c2f9db0-pxd-enc706349485201319665
+ * # cat /sys/dev/block/253:0/dm/name
+ *    pxd-enc706349485201319665
+ */
+int is_pxd_enc_device(int major, int minor)
+{
+    char filename[256] = {0};
+    char pxd_namestr[10] = {0};
+    char *pattern = "pxd-enc\0";
+    sprintf(filename, "/sys/dev/block/%d:%d/dm/name", major, minor);
+    read_to_str(filename, pxd_namestr, 9);
+    if (strlen(pxd_namestr) > 0)
+    {
+        if (strncmp(pxd_namestr, pattern, strlen(pattern)) == 0)
+        {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/*
+ * For loopback devices
+ * # cat /sys/dev/block/7:0/loop/backing_file
+ *    /dev/pxd/pxd7871666956677911
+ */
+int is_pxd_passthrough_device(int major, int minor)
+{
+    char filename[256] = {0};
+    char pxd_namestr[20] = {0};
+    char *pattern = "/dev/pxd/pxd\0";
+    sprintf(filename, "/sys/dev/block/%d:%d/loop/backing_file", major, minor);
+    read_to_str(filename, pxd_namestr, 19);
+    if (strlen(pxd_namestr) > 0)
+    {
+        if (strncmp(pxd_namestr, pattern, strlen(pattern)) == 0)
+        {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+int is_px_device(int major, int minor)
+{
+    int is_true;
+
+    is_true = is_pxd_reg_device(major, minor);
+    if (is_true)
+        return 1;
+
+    is_true = is_pxd_enc_device(major, minor);
+    if (is_true)
+        return 1;
+
+    is_true = is_pxd_passthrough_device(major, minor);
+    if (is_true)
+        return 1;
+
+    return 0;
+}
+
+int get_major_minor_nr(const char *path, int *major, int *minor)
+{
+    struct stat statbuf = {0};
+    int rc = 0;
+    rc = lstat(path, &statbuf);
+    if (rc != 0)
+    {
+        *major = 0;
+        *minor = 0;
+        return -1;
+    }
+    unsigned int maj = (unsigned int)((statbuf.st_dev & 0x000000000000ff00) >> 8);
+    unsigned int min = (unsigned int)((statbuf.st_dev & 0x00000000000000ff));
+    *major = maj;
+    *minor = min;
+    return 0;
+}
+
+int statfs(const char *path, struct statfs *buf)
+{
+    int rc = -1;
+
+    // Perform the actual system call
+    rc = real_statfs(path, buf);
+
+    if (rc == 0)
+    {
+        // TODO: do we need to check buf->f_type in (ext4 | XFS) ?
+        int major = 0;
+        int minor = 0;
+        int rc2 = get_major_minor_nr(path, &major, &minor);
+        if (rc2 != 0)
+        {
+            return rc;
+        }
+
+        int is_pxd = is_px_device(major, minor);
+        if (is_pxd == 1)
+        {
+            buf->f_type = 0x6969; // NFS_SUPER_MAGIC
```

Comment:
Is the incoming argument always a shared mountpath?

Reply:
Kubevirt has a check that prevents "live migration" if the VM is not using a shared volume. Therefore, we should not get here.

In addition, I don't think we allow mounting a non-shared volume on 2 hosts at the same time. Therefore, I am guessing that this is safe for non-shared volumes as well.